db/changes.go

```diff
@@ -87,31 +87,52 @@ func (db *Database) addDocToChangeEntry(entry *ChangeEntry, options ChangesOptio
 		return
 	}
 
-	var doc *document
-	var err error
-	// The document, which may include just the syncMeta or may include syncMeta + Body, depending on circumstances
-	if options.IncludeDocs {
-		// load whole doc
-		doc, err = db.GetDocument(entry.ID, DocUnmarshalAll)
+	// Three options for retrieving document content, depending on what's required:
+	//   includeConflicts only:
+	//      - Retrieve document metadata from bucket (required to identify current set of conflicts)
+	//   includeDocs only:
+	//      - Use rev cache to retrieve document body
+	//   includeConflicts and includeDocs:
+	//      - Retrieve document AND metadata from bucket; single round-trip usually more efficient than
+	//      metadata retrieval + rev cache retrieval (since rev cache miss will trigger KV retrieval of doc+metadata again)
+
+	if options.IncludeDocs && includeConflicts {
+		// Load doc body + metadata
+		doc, err := db.GetDocument(entry.ID, DocUnmarshalAll)
 		if err != nil {
 			base.Warn("Changes feed: error getting doc %q: %v", entry.ID, err)
 			return
 		}
+		db.AddDocInstanceToChangeEntry(entry, doc, options)
 
-	} else {
-		// get doc metadata
-		doc = &document{}
+	} else if includeConflicts {
+		// Load doc metadata only
+		doc := &document{}
+		var err error
 		doc.syncData, err = db.GetDocSyncData(entry.ID)
 		if err != nil {
 			base.Warn("Changes feed: error getting doc sync data %q: %v", entry.ID, err)
 			return
 		}
+		db.AddDocInstanceToChangeEntry(entry, doc, options)
 
+	} else if options.IncludeDocs {
+		// Retrieve document from rev cache
+		revID := entry.Changes[0]["rev"]
+		err := db.AddDocToChangeEntryUsingRevCache(entry, revID)
+		if err != nil {
+			base.Warn("Changes feed: error getting revision body for %q (%s): %v", entry.ID, revID, err)
```

Comment:
What about adding some more explanation about the implications of this?  It might seem obvious, but I think even just adding something like "Omitting this revision from the changes feed, cblite clients will not receive this revision".  

I think making these warnings easier to interpret and act upon would increase the supportability.  /cc @JFlath for input.

Reply:
I think we can add a bit more information, but it's important not to make unjustified assumptions about the request context.  e.g. here we don't know whether the changes request is being made by CBL, and it's only the revision body that won't be included in the changes feed, not the revision itself.

Reply:
> here we don't know whether the changes request is being made by CBL

Fair enough

> and it's only the revision body that won't be included in the changes feed, not the revision itself

I think that would be a great thing to include.  If I were reading the logs, I would want to see something like:

> "Warning: partial error generating changes feed.  Unable to include revision body in response for doc %doc_id /  revision %rev_id due to error: %err.  Sync Gateway clients will need to explicitly re-request this revision body, otherwise will not receive it.

Reply:
I agree that we want more self-supporting/describing error messages, but don't think there's value in comments like "Sync Gateway clients will need to explicitly re-request this revision body, otherwise will not receive it.", as it doesn't really provide any insight on what exactly is going on, or what clients should be doing differently (e.g. why would the revision be available on a re-request).  I think we should strive for a clear description of why the warning is being raised, and avoid prescriptive comments unless we know they definitely apply.

Reply:
> as it doesn't really provide any insight on what exactly is going on

Putting myself in the position of someone running sync gateway, I guess I would care about the following things:

- Severity / implications.  Eg, the class of downstream failures it could cause, so I know how to prioritize this.  
- Possible causes, so I know what I (possibly) did to cause this and what I could do to prevent this.

I guess a good metric for an error message would be whether a user sees the warning and feels the need to file a ticket to find out whether it's causing issue X or not.  

In the Couchbase Lite case, what are the implications of a client not getting a revision body on the changes feed?  Will it retry and eventually get it, or will it fail to ever get it?  Per your point though, it might not be a Couchbase Lite client requesting this changes feed, which makes it particularly tricky to describe accurately in 100% of the cases.

If it's impossible to explain the error concisely in a sentence, it might make sense to look into using unique error codes that are linked to documentation with verbose explanations.

Reply:
I think this is better discussed holistically for Sync Gateway, particularly around distinctions between Warning and Info logging.  I don't agree with all of your comments above, but I agree the discussion makes sense to have.  I'm not going to make any changes in the context of this PR, though.